noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -1,76 +1,120 @@
     #[private]
-    fn fee_entrypoint_public(amount: Field, asset: AztecAddress, nonce: Field) {
-        FPC::at(context.this_address())
-            .prepare_fee(context.msg_sender(), amount, asset, nonce)
+    fn fee_entrypoint_public(max_fee: Field, nonce: Field) {
+        // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
+        let config = storage.config.read();
+
+        // We pull the max fee from the user's balance of the accepted asset to this contract.
+        // docs:start:public_call
+        Token::at(config.accepted_asset)
+            .transfer_in_public(context.msg_sender(), context.this_address(), max_fee, nonce)
             .enqueue(&mut context);
+        // docs:end:public_call
+
         context.set_as_fee_payer();
         // TODO(#6277) for improving interface:
-        // FPC::at(context.this_address()).pay_refund(context.msg_sender(), amount, asset).set_public_teardown_function(&mut context);
+        // FPC::at(context.this_address()).pay_refund(...).set_public_teardown_function(&mut context);
         context.set_public_teardown_function(
             context.this_address(),
             comptime { FunctionSelector::from_signature("pay_refund((Field),Field,(Field))") },
-            [context.msg_sender().to_field(), amount, asset.to_field()],
+            [context.msg_sender().to_field(), max_fee, config.accepted_asset.to_field()],
         );
     }
 
+    /// Pays the refund to the `refund_recipient`. The refund is the difference between the `max_fee` and
+    /// the actual fee. `accepted_asset` is the asset in which the refund is paid. It's passed as an argument
+    /// to avoid the need for another read from public storage.
     #[public]
     #[internal]
-    fn prepare_fee(from: AztecAddress, amount: Field, asset: AztecAddress, nonce: Field) {
-        // docs:start:public_call
-        Token::at(asset).transfer_in_public(from, context.this_address(), amount, nonce).call(
-            &mut context,
-        );
-        // docs:end:public_call
+    fn pay_refund(refund_recipient: AztecAddress, max_fee: Field, accepted_asset: AztecAddress) {
+        let actual_fee = context.transaction_fee();
+        assert(!max_fee.lt(actual_fee), "Max fee paid to the paymaster does not cover actual fee");
+        let refund = max_fee - actual_fee;
+
+        Token::at(accepted_asset)
+            .transfer_in_public(context.this_address(), refund_recipient, refund, 0)
+            .call(&mut context);
     }
 
     #[public]
-    #[internal]
-    fn pay_refund(refund_address: AztecAddress, amount: Field, asset: AztecAddress) {
-        // Just do public refunds for the present
-        let refund = compute_rebate(context, amount);
-        Token::at(asset).transfer_in_public(context.this_address(), refund_address, refund, 0).call(
-            &mut context,
-        );
+    #[view]
+    fn get_accepted_asset() -> AztecAddress {
```

Comment:
Exposed this so that we can load this info directly from contract instead of having to pass that value in our TS API.